genesis/engine/solvers/fem_solver.py

```diff
@@ -1317,3 +1398,11 @@ def n_elements_max(self):
     @property
     def vol_scale(self):
         return self._vol_scale
+
+    @property
+    def n_surface_vertices(self):
+        return self.surface_vertices_np.shape[0]
```

Comment:
Compute this property dynamically based on the shape of taichi field `self.vertices_on_surface` directly (without doing any conversion to either torch or numpy). If this is computationally costly, then use caching to compute the value only once during build and return the pre-computed value ever after.